File: ex33_search_rotated_array/solution.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if not nums:
            return -1
        
        pivot = self.find_pivot(nums)
        if target == nums[pivot]:
            return pivot
        
        lower_search = self.binary_search(nums[pivot:], target)
        if lower_search != -1:
            return pivot + lower_search
        
        upper_search = self.binary_search(nums[:pivot], target)
        if upper_search != -1:
            return upper_search 
        
        return -1
    
    def find_pivot(self, l):
        left, right = 0, len(l) - 1
        
        while left < right:
            mid = (left + right) / 2
            if mid == left:
                break
                
            if l[mid] < l[left]:
                right = mid
            else:
                left = mid
                
        return right if l[right] < l[left] else left
                
    def binary_search(self, l, target):
        left, right = 0, len(l) - 1
        
        while left <= right:
            
            mid = (left + right) / 2
            if l[mid] == target:
                return mid
            elif l[mid] < target:
                left = mid + 1
            elif l[mid] > target:
                right = mid - 1
                
        return -1

            
            

File: ex33_search_rotated_array/solution.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if (nums.empty())
            return -1;
        
        size_t pivot = findPivot(nums);
        int lowerIndex = binarySearchRange(nums, pivot, nums.size() - 1, target);
        if (lowerIndex != -1) {
            return lowerIndex;
        }
        
        int upperIndex = binarySearchRange(nums, 0, pivot, target);
        if (upperIndex != -1) {
            return upperIndex;
        }
        
        return -1;
    }
    
private:
    size_t findPivot(const vector<int>& nums) {
        size_t left = 0, right = nums.size() - 1;
        while (left < right) {
            size_t mid = (left + right) / 2;
            if (mid == left)
                break;
            
            if (nums[mid] < nums[left]) {
                right = mid;       
            } else {
                left = mid;
            }
        }
        return (nums[left] < nums[right]) ? left : right;
    }
    
    int binarySearchRange(const vector<int>& nums, int left, int right, const int target) {
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return -1;
    }
};
